src/runtime/HeapValue.h

```diff
@@ -0,0 +1,515 @@
+                    return;
+                }
+            }
+            m_data.payload = reinterpret_cast<intptr_t>(new DoubleInHeapValue(from.asNumber()));
+            return;
+        }
+
+#ifdef ESCARGOT_32
+        m_data.payload = ~from.tag();
+#else
+        m_data.payload = from.payload();
+#endif
+    }
+
+protected:
+    void fromValueForCtor(const Value& from)
+    {
+        if (from.isPointerValue()) {
+#ifdef ESCARGOT_32
+            ASSERT(!from.isEmpty());
+#endif
+            m_data.payload = (intptr_t)from.asPointerValue();
+        } else {
+            int32_t i32;
+            if (from.isInt32() && HeapValueImpl::PlatformSmiTagging::IsValidSmi(i32 = from.asInt32())) {
+                m_data.payload = HeapValueImpl::PlatformSmiTagging::IntToSmi(i32);
+            } else if (from.isNumber()) {
+                m_data.payload = reinterpret_cast<intptr_t>(new DoubleInHeapValue(from.asNumber()));
+            } else {
+#ifdef ESCARGOT_32
+                m_data.payload = ~from.tag();
+#else
+                m_data.payload = from.payload();
+#endif
+            }
+        }
+    }
+
+    explicit HeapValue(HeapValueData v)
+        : m_data(v)
+    {
+    }
+
+private:
+    HeapValueData m_data;
+};
+
+#if defined(ESCARGOT_USE_32BIT_IN_64BIT)
+class HeapSmallValue {
+public:
+    enum EmptyValueInitTag { EmptyValue };
+
+    HeapSmallValue(EmptyValueInitTag)
+    {
+        m_data.payload = (intptr_t)(ValueEmpty);
+    }
+
+    HeapSmallValue()
+    {
+        m_data.payload = (intptr_t)(ValueUndefined);
+    }
+
+    HeapSmallValue(const HeapSmallValue& from)
+    {
+        m_data.payload = from.m_data.payload;
+    }
+
+    HeapSmallValue(const Value& from)
+    {
+        if (from.isPointerValue()) {
+            m_data = HeapSmallValueData(from.asPointerValue());
+        } else {
+            int32_t i32;
+            if (from.isInt32() && HeapValueImpl::PlatformSmiTagging::IsValidSmi(i32 = from.asInt32())) {
+                m_data.payload = HeapValueImpl::PlatformSmiTagging::IntToSmi(i32);
+            } else if (from.isNumber()) {
+                m_data = HeapSmallValueData(new DoubleInHeapValue(from.asNumber()));
+            } else {
+                m_data.payload = from.payload();
+            }
+        }
+    }
+
+    HeapSmallValue(const HeapValue& from)
+    {
+        ASSERT(from.payload() <= std::numeric_limits<uint32_t>::max());
+        m_data.payload = from.payload();
+    }
+
+    ALWAYS_INLINE operator HeapValue() const
+    {
+        return HeapValue::fromPayload(reinterpret_cast<void*>(m_data.payload));
+    }
+
+    static HeapSmallValue fromPayload(void* p)
+    {
+        HeapSmallValue v;
+        v.m_data = HeapSmallValueData(p);
+        return v;
+    }
+
+    uint32_t payload() const
+    {
+        return m_data.payload;
+    }
+
+    ALWAYS_INLINE operator Value() const
+    {
+        if (HAS_SMI_TAG(m_data.payload)) {
+            int32_t value = (int)(m_data.payload) >> 1;
```

Comment:
I tried to use that function, but we cannot use it.
the function casts value as intptr_t. when the value has negaive value,
the value is converted to positive value because size is differ(64 vs 32)